KAFKA-4818: Exactly once transactional clients

apurvam opened the pull request.

apurvam committed:
CPKAFKA-465 : Implement READ_COMMITTED mode in the KafkaConsumer (#145)

Commit:
Complete implementation of the transactional producer.

Commit:
Drop the consumerCoordinator field from the AddOffsetsToTxn response.
Implemented the ability to find the group coordinator from the producer.

Commit:
Add break in FindCoordinatorCallback

ijuma commented on checkstyle/checkstyle.xml:
```diff
@@ -102,7 +102,7 @@
     <module name="ClassDataAbstractionCoupling">
       <!-- default is 7 -->
-      <property name="max" value="15"/>
+      <property name="max" value="22"/>
```
Hmm, if we keep increasing this, then not sure there's any point in having it. Can we exclude the offending classes individually? Same for the CyclomaticComplexity change.

ijuma commented on checkstyle/suppressions.xml:
```diff
@@ -47,6 +47,8 @@
               files="Fetcher.java"/>
     <suppress checks="CyclomaticComplexity"
+              files="Sender.java"/>
+    <suppress checks="CyclomaticComplexity"
```
We can use `|` like we are doing for ClassFanOutComplexity`. Same for `NPathComplexity`.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -234,6 +235,17 @@
+            " which have been aborted. Non-transactional messages will be returned unconditionally in either mode.</p> <p>Messages will be  always returned in offset order. Hence, in " +
+            " <code>READ_COMMITTED</code> mode, consumer.poll() will only return messages upto the last resolved (committed or aborted) transaction. In particular any messages appearing after" +
+            " messages belonging onging transactions will be withheld until the said transaction has been completed and its messages are delivered to the application. As a result, <code>READ_COMMITTED</code>" +
+            " consumers will not be able to read upto the log end offset when there are inflight transactions.</p>";
```
`Messages will be  always returned` should probably be `Messages will always be returned`, `onging` -> `to ongoing`, `until the said transaction` -> `until the relevant transaction`, `upto` -> `up to`.

It's a bit unclear what "and its messages are delivered to the application" refers to here. Does it add anything or shall we just exclude it?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -405,6 +417,11 @@
+                                .define(ISOLATION_LEVEL_CONFIG,
+                                        Type.STRING,
+                                        DEFAULT_ISOLATION_LEVEL,
+                                        Importance.LOW,
```
Is this really `LOW`? Seems like `MEDIUM` at least. Or is it because this is a dev preview? If so, we may want to include mention of that in the config docs.

apurvam replied:
I changed it to `MEDIUM`. I am not sure how to choose between priorities.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -660,6 +662,10 @@ private KafkaConsumer(ConsumerConfig config,
             ChannelBuilder channelBuilder = ClientUtils.createChannelBuilder(config);
+
+            this.isolationLevel = IsolationLevel.valueOf(
+                    config.getString(ConsumerConfig.ISOLATION_LEVEL_CONFIG).toUpperCase(Locale.ENGLISH));
```
`Locale.ROOT` is what we typically use.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -710,7 +716,8 @@ private KafkaConsumer(ConsumerConfig config,
                     this.time,
-                    this.retryBackoffMs);
+                    this.retryBackoffMs,
+                    IsolationLevel.READ_UNCOMMITTED);
```
We should use the variable here instead of hardcoding it, right?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -532,6 +533,7 @@
     private final Metadata metadata;
     private final long retryBackoffMs;
     private final long requestTimeoutMs;
+    private final IsolationLevel isolationLevel;
```
It doesn't look like we need this field.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -738,7 +745,8 @@ private KafkaConsumer(ConsumerConfig config,
                   long retryBackoffMs,
-                  long requestTimeoutMs) {
+                  long requestTimeoutMs,
+                  String isolationLevel) {
```
It doesn't look like we need this parameter.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -440,7 +451,8 @@ else if (strategy == OffsetResetStrategy.LATEST)
 
     public Map<TopicPartition, Long> endOffsets(Collection<TopicPartition> partitions, long timeout) {
-        return beginningOrEndOffset(partitions, ListOffsetRequest.LATEST_TIMESTAMP, timeout);
+        long endTimestamp = isolationLevel == IsolationLevel.READ_COMMITTED ? ListOffsetRequest.LSO_TIMESTAMP : ListOffsetRequest.LATEST_TIMESTAMP;
```
I suggest we extract a method `endTimestamp` or something and use it from `resetOffset` and `endOffset`

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -919,8 +941,10 @@ private void maybeEnsureValid(Record record) {
-            if (records != null)
+            if (records != null) {
                 records.close();
+                records = null;
```
Out of curiosity, was this causing issues?

apurvam replied:
No. I think I put it in when debugging a test failure and then forgot to remove it.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -431,6 +444,83 @@ private static int parseAcks(String acksString) {
+     */
+    public void beginTransaction() throws ProducerFencedException {
+        // Set the transactional bit in the producer.
+        assert transactionState != null && transactionState.isTransactional() && !transactionState.isInTransaction() && transactionState.hasPid();
```
We don't usually use JVM level asserts because they are disabled by default. Same for all other cases in this PR.

apurvam replied:
Right. I had that in there as a convenience during testing. Changed to throwing `IllegalStateException` if the conditions are not met.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java:
```diff
@@ -96,6 +98,71 @@ public MockProducer(boolean autoComplete, Partitioner partitioner, Serializer<K>
+     *         in the configuration.
+     */
+    public void initTransactions() {
+
```
Should we hold some state indicating that this was invoked and then throw the relevant exception if we don't invoke it when we should? Same for `beginTransaction`.

apurvam replied:
We actually do this in the real producer. Or do you mean for the mock producer as well?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java:
```diff
@@ -96,6 +98,71 @@ public MockProducer(boolean autoComplete, Partitioner partitioner, Serializer<K>
+     *
+     * @throws IllegalStateException if the TransactionalId for the producer is not set
+     *         in the configuration.
+     */
```
I don't think we need the docs in the `MockProducer`. It's enough to have them in `KafkaProducer`.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java:
```diff
@@ -96,6 +98,71 @@ public MockProducer(boolean autoComplete, Partitioner partitioner, Serializer<K>
+     * @throws ProducerFencedException if another producer is with the same
+     *         transactional.id is active.
+     */
+    public void sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets,
```
Hmm, it would be good to have some implementation for all the new methods.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/Producer.java:
```diff
@@ -35,6 +37,61 @@
+    /**
+     * Should be called before the start of each new transaction.
+     *
+     * @throws ProducerFencedException if another producer is with the same
```
`producer is with` -> `producer with`

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -325,7 +337,17 @@
+                                        TRANSACTION_TIMEOUT_DOC)
+                                .define(TRANSACTIONAL_ID_CONFIG,
+                                        Type.STRING,
+                                        "",
```
It's a bit clearer if the default is `null` instead of empty string.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -325,7 +337,17 @@
+                                .define(TRANSACTIONAL_ID_CONFIG,
+                                        Type.STRING,
+                                        "",
+                                        Importance.LOW,
```
Should this be `LOW` or `MEDIUM`?

apurvam replied:
Again, How do we decide between the two?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -47,32 +185,166 @@ public boolean isValid() {
+            case TRANSACTION:
+                return transactionCoordinator;
+            default:
+                return null;
```
It sucks that Java doesn't have exhaustiveness warnings for enums. Given that, it seems that throwing an assertion error is the best bet for impossible paths like this.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -47,32 +185,166 @@ public boolean isValid() {
+    }
+
+    public void resetInFlightRequestCorrelationId() {
+        inFlightRequestCorrelationId = Integer.MIN_VALUE;
```
We should probably just use `-1` for the unset value? Or `null`. The latter is a little safer, but either option seems easier to recognise than a really large negative number.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -108,6 +378,15 @@ public synchronized void setPidAndEpoch(long pid, short epoch) {
      * messages will return an OutOfOrderSequenceException.
      */
     public synchronized void resetProducerId() {
+        if (isTransactional()) {
```
In the javadoc above, `it's` -> `its`. Also, don't we have to update it given the change to the implementation?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -542,6 +545,17 @@ public void awaitFlushCompletion() throws InterruptedException {
+    public boolean hasUnflushedBatches() {
+        boolean hasUnflushed = false;
+        for (Map.Entry<TopicPartition, Deque<ProducerBatch>> entry : this.batches().entrySet()) {
+            if (0 < entry.getValue().size()) {
```
`if (!entry.getValue().isEmpty())`

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -542,6 +545,17 @@ public void awaitFlushCompletion() throws InterruptedException {
+        for (Map.Entry<TopicPartition, Deque<ProducerBatch>> entry : this.batches().entrySet()) {
+            if (0 < entry.getValue().size()) {
+                hasUnflushed = true;
+                break;
```
I think we can just say `return true` here.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -542,6 +545,17 @@ public void awaitFlushCompletion() throws InterruptedException {
+                break;
+            }
+        }
+        return 0 < this.incomplete.incomplete.size() || hasUnflushed;
```
`return !this.incomplete.incomplete.isEmpty` if we make the other change.

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:
```diff
@@ -255,6 +242,20 @@ public void close() {
+
+            if (magic < RecordBatch.MAGIC_VALUE_V2)
+                throw new IllegalArgumentException("Idempotent messages are not supported for magic " + magic);
+        }
```
We should probably do these checks before `closeForRecordAppends` right?

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java:
```diff
@@ -55,24 +55,34 @@
         private final short acks;
         private final int timeout;
         private final Map<TopicPartition, MemoryRecords> partitionRecords;
+        private final String transactioanlId;
```
Typo.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java:
```diff
@@ -83,6 +93,7 @@ public String toString() {
                     .append(", acks=").append(acks)
                     .append(", timeout=").append(timeout)
                     .append(", partitionRecords=(").append(partitionRecords)
+                    .append(", transactionalId=(").append(transactioanlId != null ? transactioanlId : "")
```
Seems like we need to add a closing `)` to the previous line.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java:
```diff
@@ -100,12 +111,15 @@ public String toString() {
     private volatile Map<TopicPartition, MemoryRecords> partitionRecords;
 
     private ProduceRequest(short version, short acks, int timeout, Map<TopicPartition, MemoryRecords> partitionRecords) {
```
Is this used by anything? If not, we should remove it.

ijuma commented on core/src/test/scala/unit/kafka/server/FetchRequestTest.scala:
```diff
@@ -156,8 +156,7 @@ class FetchRequestTest extends BaseRequestTest {
       "key", new String(new Array[Byte](maxPartitionBytes + 1)))).get
-    val fetchRequest = FetchRequest.Builder.forConsumer(Int.MaxValue, 0,
-      createPartitionMap(maxPartitionBytes, Seq(topicPartition))).build(2)
+    val fetchRequest = FetchRequest.Builder.forConsumer(Int.MaxValue, 0, createPartitionMap(maxPartitionBytes, Seq(topicPartition))).build(2)
```
This line is a bit longer than what we allow.

ijuma reviewed (commented):
Thanks for the PR. Did a first pass and left some comments.

Since we are missing integration tests given that the server-side is not ready, it would be good to file a JIRA about that.

Commit:
Addressed PR comments

Commit:
Fix FindBugs errors

apurvam commented:
Thanks for the review @ijuma!

For some context. The main changes for the producer are in `Sender` and in `TransactionState`. The `TransactionState` class has nearly all the code for handling transactions. 

Essentially, the producer continues to have a single background thread it uses to send requests. The main difference is that transactional RPCs have the highest priority: if there is any such request waiting to be sent or in flight, no `ProduceRequest`s are sent. 

The `TransactionState` maintains a queue of Transactional requests, ordered by priority. Some requests like `FindCoordinator` must always be sent first. Also `EndTxn` must always go last, etc. This class also manages chaining requests: so if we need to send to a coordinator, but don't have a coordinator yet, it will put the `FindCoordinator` request at the head of the queue. 

The `TransactionsTest` has all the tests for the new flows.

On the consumer side, things are much simpler. All the main logic is in the `Fetcher`  in the `PartitionRecords` sub class. Here we decide whether to drop batches based on the contents of the abort index received in the fetch request. 

The `FetcherTest` has new test cases for the consumer side fetcher changes.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -234,6 +235,17 @@
+            " which have been aborted. Non-transactional messages will be returned unconditionally in either mode.</p> <p>Messages will always be returned in offset order. Hence, in " +
+            " <code>READ_COMMITTED</code> mode, consumer.poll() will only return messages upto the last resolved (committed or aborted) transaction. In particular any messages appearing after" +
+            " messages belonging to onging transactions will be withheld until the relevant transaction has been completed. As a result, <code>READ_COMMITTED</code>" +
+            " consumers will not be able to read up to the log end offset when there are inflight transactions.</p>";
```
nit: consumers read up to the high watermark, not the log end offset.

Also, it might be worth mentioning something about how READ_COMMITTED makes use of the LSO in APIs like `seekToEnd`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -234,6 +235,17 @@
+            " transactional messages which have been committed. If set to <code>READ_UNCOMMITTED</code>' (the default), consumer.poll() will return all messages, even transactional messages" +
+            " which have been aborted. Non-transactional messages will be returned unconditionally in either mode.</p> <p>Messages will always be returned in offset order. Hence, in " +
+            " <code>READ_COMMITTED</code> mode, consumer.poll() will only return messages upto the last resolved (committed or aborted) transaction. In particular any messages appearing after" +
+            " messages belonging to onging transactions will be withheld until the relevant transaction has been completed. As a result, <code>READ_COMMITTED</code>" +
```
typo: "onging"

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -234,6 +235,17 @@
+    public static final String ISOLATION_LEVEL_DOC = "<p>Controls how to read messages written transactionally. If set to <code>READ_COMMITTED</code>, consumer.poll() will only return" +
+            " transactional messages which have been committed. If set to <code>READ_UNCOMMITTED</code>' (the default), consumer.poll() will return all messages, even transactional messages" +
+            " which have been aborted. Non-transactional messages will be returned unconditionally in either mode.</p> <p>Messages will always be returned in offset order. Hence, in " +
+            " <code>READ_COMMITTED</code> mode, consumer.poll() will only return messages upto the last resolved (committed or aborted) transaction. In particular any messages appearing after" +
```
nit: "up to"

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -405,6 +417,11 @@
                                                 Type.BOOLEAN,
                                                 true,
                                                 Importance.LOW)
+                                .define(ISOLATION_LEVEL_CONFIG,
```
We can restrict the allowable values here. See the use of the auto reset definition for an example.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -234,6 +235,17 @@
 
+    /** <code>isolation.level</code> */
+    public static final String ISOLATION_LEVEL_CONFIG = "isolation.level";
+    public static final String ISOLATION_LEVEL_DOC = "<p>Controls how to read messages written transactionally. If set to <code>READ_COMMITTED</code>, consumer.poll() will only return" +
```
Wondering a bit about the config values that we should use. Would "committed" and "uncommitted" be clear enough?

By the way, I'm also realizing there may be some confusion with the use of the term "committed." I think historically we have considered data to be "committed" when it has been properly replicated, but we now mean more than that. Not sure what we can do about it other than mention it in the documentation.

apurvam replied:
I changed everything to lower case, but retained the 'read' prefix, since we are trying to use the database terminology with 'isolation.level'.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -440,7 +451,11 @@ else if (strategy == OffsetResetStrategy.LATEST)
+    }
+
+    private long endTimestamp() {
+        return isolationLevel == IsolationLevel.READ_COMMITTED ? ListOffsetRequest.LSO_TIMESTAMP : ListOffsetRequest.LATEST_TIMESTAMP;
```
Discussed offline, but since we've decided to take a slightly different route (i.e. passing through isolation level in the ListOffsets request), maybe we can just remove this code? It won't do anything until the server-side implementation is there anyway.

ijuma replied:
Are we going to bump the version of `ListOffsetsRequest` again then?

hachikuji replied:
Yes, a version bump was included in the KIP proposal, though the details will be a little different.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -865,6 +886,8 @@ public void onAssignment(Set<TopicPartition> assignment) {
         private final TopicPartition partition;
         private final CompletedFetch completedFetch;
         private final Iterator<? extends RecordBatch> batches;
+        private final Set<Long> abortedPids;
```
nit: it's a bit more verbose, but we've preferred in other places spelling out `producerId` rather than using the abbreviation.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -933,6 +960,10 @@ private Record nextFetchedRecord() {
                         return null;
                     }
                     currentBatch = batches.next();
+                    if (isBatchAborted(currentBatch, abortedPids, abortedTransactions)) {
```
Should this go after the call to `maybeEnsureValid`? I'm wondering if a pathological data corruption might cause us to mistakenly skip over data.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
             return records;
         }
+
+        private boolean isBatchAborted(RecordBatch batch, Set<Long> abortedPids, Queue<FetchResponse.AbortedTransaction> abortedTransactions) {
```
nit: do we need to pass `abortedPids` and `abortedTransactions` as arguments since they are already class fields?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
+                        abortedPids.add(batch.producerId());
+                        abortedTransactions.remove(nextAbortedTransaction);
+                    }
+                    log.trace("Skipping aborted record with pid {} and base offset {}", batch.producerId(), batch.baseOffset());
```
nit: "Skipping aborted record **batch** ..."? Also, spell out "producerId" and include the topic partition?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
+            *      this means that the entry has been aborted. Add the pid to the aborted pids set, and remove
+            *      the entry from the abort index.
+            */
+            if (isolationLevel == IsolationLevel.READ_COMMITTED) {
```
nit: This may be a little more intuitive if we move this check into the caller:

```java
if (isolationLevel == IsolationLevel.READ_COMMITTED && isBatchAborted(currentBatch))
```

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
+                FetchResponse.AbortedTransaction nextAbortedTransaction = abortedTransactions.peek();
+                if (abortedPids.contains(batch.producerId())
+                        || (nextAbortedTransaction != null && nextAbortedTransaction.producerId == batch.producerId() && nextAbortedTransaction.firstOffset <= batch.baseOffset())) {
+                    if (abortedPids.contains(batch.producerId()) && containsAbortMarker(batch)) {
```
It's a bit annoying that we need to check inclusion in `abortedPids` again. Is it possible to restructure this to treat that branch separately?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
+                        abortedPids.remove(batch.producerId());
+                    } else if (nextAbortedTransaction != null && nextAbortedTransaction.producerId == batch.producerId() && nextAbortedTransaction.firstOffset <= batch.baseOffset()) {
+                        abortedPids.add(batch.producerId());
+                        abortedTransactions.remove(nextAbortedTransaction);
```
Should we use `poll` instead of `remove`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
+            if (partition.abortedTransactions != null && 0 < partition.abortedTransactions.size()) {
+                abortedTransactions = new PriorityQueue<>(
+                        partition.abortedTransactions.size(),
+                        new Comparator<FetchResponse.AbortedTransaction>() {
```
Could we let `AbortedTransaction` implement `Comparable`?

ijuma replied:
One of the advantages of using `Comparator` is that it avoids issues related to consistency with `equals`.

hachikuji replied:
Yeah, that's fair.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -865,6 +886,8 @@ public void onAssignment(Set<TopicPartition> assignment) {
         private final CompletedFetch completedFetch;
         private final Iterator<? extends RecordBatch> batches;
+        private final Set<Long> abortedPids;
+        private final Queue<FetchResponse.AbortedTransaction> abortedTransactions;
```
The fact that we actually use a priority queue is significant to the correctness of the algorithm, so maybe we should not use the more generic type?

apurvam replied:
It helps ensuring that the right requests go in the right order when there are multiple transactional  requests in flight. For instance, we could have no valid coordinators, pending `AddPartitionsToTransaction` requests, and an offset commit request. In this case, we would have two `FindCoordinatorRequests` (one at a time), one or more `AddPartitionToTransactionRequest`, and a `AddOffsetsToTransactionRequest` in the queue. With the priority queue, we are guaranteed that the find coordinator requests will go first. 

Conversely, it also ensures that the `EndTransactionRequest` goes last.

hachikuji replied:
I think your comment is for the priority queue used in the producer. All I was suggesting here is that we replace `Queue` with `PriorityQueue` to emphasize that not just any queue will work.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
+            Iterator<Record> batchIterator = batch.iterator();
+            Record firstRecord = batchIterator.hasNext() ? batchIterator.next() : null;
+            if (firstRecord != null && batchIterator.hasNext())
+                throw new CorruptRecordException("A RecordBatch containing a control message contained more than one message.");
```
nit: "RecordBatch" -> "record batch"? Also, may want to include a bit more detail here such as the partition and the offset.

ijuma replied:
Hmm, where have we checked that the record batch contains a control message? The exception seems a bit out of place.

apurvam replied:
Ismael, you are correct. I think this exception itself is a bug.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
+            Record firstRecord = batchIterator.hasNext() ? batchIterator.next() : null;
+            if (firstRecord != null && batchIterator.hasNext())
+                throw new CorruptRecordException("A RecordBatch containing a control message contained more than one message.");
+            return firstRecord != null && firstRecord.isControlRecord() && ControlRecordType.ABORT == ControlRecordType.parse(firstRecord.key());
```
Wondering if we should add a `controlRecordType()` method to the `Record` interface?

ijuma replied:
Would it make sense to add `RecordBatch.containsControlRecordType`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
+
+        private Queue<FetchResponse.AbortedTransaction> abortedTransactions(FetchResponse.PartitionData partition) {
+            PriorityQueue<FetchResponse.AbortedTransaction> abortedTransactions = null;
+            if (partition.abortedTransactions != null && 0 < partition.abortedTransactions.size()) {
```
nit: replace `0 < partition.abortedTransactions.size()` with `!partition.abortedTransactions.isEmpty()`.

hachikuji reviewed (commented):
Some comments on the consumer side.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/Producer.java:
```diff
@@ -35,6 +37,61 @@
+     * @throws IllegalStateException if the TransactionalId for the producer is not set
+     *         in the configuration.
+     */
+    void initTransactions() throws IllegalStateException;
```
nit: no need for the `throws` here.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/Producer.java:
```diff
@@ -35,6 +37,61 @@
 public interface Producer<K, V> extends Closeable {
 
     /**
+     * Needs to be called before any of the other transaction methods. Assumes that
```
Rather than duplicating these javadocs, maybe we can just link to the `KafkaProducer` method like we inside `Consumer`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -365,11 +369,20 @@ private static int configureRequestTimeout(ProducerConfig config, Map<String, Ob
+                log.info("Instantiated a transactional producer.");
+            else
+                log.info("Instantiated an idempotent producer.");
+        } else if (config.originals().containsKey(ProducerConfig.TRANSACTIONAL_ID_CONFIG)) {
```
Perhaps if the user configures a transactionalId, then we should enable idempotence automatically. We can raise an exception only if the user has explicitly disabled idempotence.

ijuma replied:
One concern is making things too magical. However, it does seem that there's no scenario where someone would want transactions without idempotence. So, I'm in favour of this suggestion.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -431,6 +444,96 @@ private static int parseAcks(String acksString) {
+        if (transactionState == null || !transactionState.isTransactional()) {
+            throw new IllegalStateException("Cannot call initTransactions without setting a transactional id.");
+        }
+        FutureTransactionalResult result = transactionState.initializeTransactions();
```
nit: seems like there's no real need to use the specific type `FutureTransactionalResult ` instead of `Future<TransactionalRequestResult>`. Also, could we just let `TransactionalRequestResult` implement `Future` and drop `FutureTransactionalResult `?

apurvam replied:
I agree. I was just emulating the pattern of `ProduceRequestResult` and `FutureRequestMetadata`. We can simplify this one though.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -431,6 +444,96 @@ private static int parseAcks(String acksString) {
+     */
+    public void beginTransaction() throws ProducerFencedException {
+        // Set the transactional bit in the producer.
+        if (transactionState == null || !transactionState.isTransactional() || transactionState.isInTransaction()
```
Seems like we could push these some of these checks in `TransactionState.beginTransaction()`. Same for the other APIs.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -29,9 +58,123 @@
+
+    private Node transactionCoordinator;
+    private Node consumerGroupCoordinator;
+    private volatile boolean isInTransaction = false;
```
Rather than having all these variables to keep track of the state, could we create an enumeration with all of the valid states (sort of like what we do on the server side). That will make it easier to ensure that state transitions are correct.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -431,6 +444,96 @@ private static int parseAcks(String acksString) {
+     *
+     * @throws ProducerFencedException if another producer with the same
+     *         transactional.id is active.
+     */
```
Maybe we should also document the illegal states? Same for the other public methods.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -523,6 +626,21 @@ private static int parseAcks(String acksString) {
+        }
+
+        if (transactionState != null && transactionState.isInTransaction()) {
+            transactionState.maybeAddPartitionToTransaction(new TopicPartition(record.topic(), record.partition()));
```
I think we need to move this check below since we haven't assigned the partition yet. We should probably add a test case for this.

apurvam replied:
I moved the check. But the test case is harder: we currently test only at the sender level, and when the `accumulator.append` happens, we already have a partition.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -523,6 +626,21 @@ private static int parseAcks(String acksString) {
     private Future<RecordMetadata> doSend(ProducerRecord<K, V> record, Callback callback) {
+        if (transactionState != null && transactionState.isTransactional() && !transactionState.hasPid()) {
+            if (transactionState.isFenced()) {
+                throw Errors.INVALID_PRODUCER_EPOCH.exception();
```
I'm not sure we want to throw the static exception since we won't get a stack trace (at least currently).

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java:
```diff
@@ -95,6 +97,27 @@ public MockProducer(boolean autoComplete, Partitioner partitioner, Serializer<K>
         this(Cluster.empty(), autoComplete, partitioner, keySerializer, valueSerializer);
     }
 
+    public void initTransactions() {
```
Probably fine not to implement these here, but perhaps add a JIRA to make sure we get to it eventually?

ijuma replied:
Yes, I think we need to make an effort to keep our mocks in good shape.

apurvam replied:
Jira here: https://issues.apache.org/jira/browse/KAFKA-5126

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -325,7 +337,17 @@
+                                        60000,
+                                        Importance.LOW,
+                                        TRANSACTION_TIMEOUT_DOC)
+                                .define(TRANSACTIONAL_ID_CONFIG,
```
Where do we enforce that the transactionalId is not empty? Maybe we could add a `NonEmptyString` validator?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -29,9 +58,123 @@
+        sequenceNumbers = new HashMap<>();
+        this.transactionalId = transactionalId;
+        this.transactionTimeoutMs = transactionTimeoutMs;
+        this.pendingTransactionalRequests = new PriorityQueue<>(2, new Comparator<TransactionalRequest>() {
```
Why do we initialize with a capacity of 2?

hachikuji reviewed (commented):
Some initial comments in the producer code.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
+                        new Comparator<FetchResponse.AbortedTransaction>() {
+                            @Override
+                            public int compare(FetchResponse.AbortedTransaction o1, FetchResponse.AbortedTransaction o2) {
+                                return (int) o1.firstOffset - (int) o2.firstOffset;
```
`Integer.compare`?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -967,6 +998,62 @@ private Record nextFetchedRecord() {
+            if (isolationLevel == IsolationLevel.READ_COMMITTED) {
+                FetchResponse.AbortedTransaction nextAbortedTransaction = abortedTransactions.peek();
+                if (abortedPids.contains(batch.producerId())
+                        || (nextAbortedTransaction != null && nextAbortedTransaction.producerId == batch.producerId() && nextAbortedTransaction.firstOffset <= batch.baseOffset())) {
```
We could perhaps add a method to `AbortedTransaction` that takes a `RecordBatch` that does these 2 checks.

ijuma replied:
Not sure if that helps in this case, but the general idea would be to push some logic outside of Fetcher and into classes that are easier to unit test.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -431,6 +444,96 @@ private static int parseAcks(String acksString) {
+            throw new IllegalStateException("Cannot send offsets to transaction. Either transactions are not enabled, or a transaction" +
+                    " isn't active right now, or a transaction is in the process of being completed, or the producer" +
+                    " has been fenced.");
+        FutureTransactionalResult result = transactionState.sendOffsetsToTransaction(offsets, consumerGroupId);
```
Seems like we don't need this variable? Same for other similar methods.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -29,9 +58,123 @@
+        this.pendingTransactionalRequests = new PriorityQueue<>(2, new Comparator<TransactionalRequest>() {
+            @Override
+            public int compare(TransactionalRequest o1, TransactionalRequest o2) {
+                return o1.priority().priority() - o2.priority.priority();
```
Integer.compare?

ijuma reviewed (commented):
A few more comments.

ijuma commented:
Also, there is a merge conflict due to https://github.com/apache/kafka/commit/17ce2a7307222d7476eb60318fab2e672eebe559

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -29,9 +58,123 @@
+        private final FindCoordinatorRequest.CoordinatorType coordinatorType;
+        private final String coordinatorKey;
+        private final RequestCompletionHandler handler;
+        // We use the priority to determine the order in which requests need to be sent out. For instance, if we have
```
Discussed briefly offline, but one of the confusing aspects of this approach is that it seems to suggest that any combination of these requests could be pending at any time, which isn't actually the case. For example, we would always flush any pending requests before sending EndTxn, so the only request type that may be pending at the same time is the FindCoordinator request. It would be nice if the approach made invariants such as this explicit. It's probably fine for a first pass, but perhaps we could add a comment indicating which combinations of requests can actually be pending at any given time. My gut feeling is that there is room for simplification once we have a clear representation of the state machine and the valid transitions.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -47,32 +190,166 @@ public boolean isValid() {
+    }
+
+    public synchronized FutureTransactionalResult sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets,
+                                         String consumerGroupId) {
```
nit: alignment

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -47,32 +190,166 @@ public boolean isValid() {
-        this.pidAndEpoch = new PidAndEpoch(NO_PRODUCER_ID, NO_PRODUCER_EPOCH);
-        this.sequenceNumbers = new HashMap<>();
-        this.time = time;
+    public boolean hasPendingTransactionalRequests() {
```
Seems this needs to be `synchronized` because the two collections it accesses are not thread-safe. Same for `nextTransactionalRequest` below.

apurvam replied:
Both of these should only be accessed from the background sender thread, so there should be no contention.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -47,32 +190,166 @@ public boolean isValid() {
+        pendingTransactionalRequests.add(request);
+    }
+
+    public Node coordinator(FindCoordinatorRequest.CoordinatorType type) {
```
nit: might be simpler to just have two methods: `transactionCoordinator()` and `consumerCoordinator()`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -47,32 +190,166 @@ public boolean isValid() {
+        newPartitionsToBeAddedToTransaction.add(topicPartition);
+    }
+
+    public void needsRetry(TransactionalRequest request) {
```
Missing synchronization? Below `didNotSend` as well.

apurvam replied:
As with the others, these are only called in the context of the background thread (in particular from `Sender.maybeSendTransactionalRequest()`), hence no need for synchronization.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -106,8 +381,15 @@ public synchronized void setPidAndEpoch(long pid, short epoch) {
      */
     public synchronized void resetProducerId() {
+        if (isTransactional())
+            return;
```
Silent `return` here seems a bit dangerous. Maybe we should throw an exception and require the caller to check  `isTransactional` itself.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -132,4 +414,332 @@ public synchronized void incrementSequenceNumber(TopicPartition topicPartition,
+        @Override
+        public void onComplete(ClientResponse response) {
+            if (response.requestHeader().correlationId() != inFlightRequestCorrelationId)
+                throw new IllegalStateException("Cannot have more than one transactional request in flight.");
```
Exceptions thrown from the completion handler will be caught by `NetworkClient`. Is the intent to propagate this back to the user? We probably need to set the error in the result and move error checking outside.

Commit:
Fix correctness issues in Fetcher and KafkaProducer which were pointed out in the PR

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -132,4 +414,332 @@ public synchronized void incrementSequenceNumber(TopicPartition topicPartition,
+            }
+        }
+
+        public abstract void handleResponse(AbstractResponse responseBody);
```
Since we're just casting the response object anyway, we could make `TransactionalRequestCallback` generic and let the type indicate the specific response type. In other words:
```java
private abstract class TransactionalRequestCallBack<T extends AbstractResponse> implements RequestCompletionHandler {
  abstract void handleResponse(T response);
}

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -132,4 +414,332 @@ public synchronized void incrementSequenceNumber(TopicPartition topicPartition,
+                throw error.exception();
+            } else if (error == Errors.INVALID_PRODUCER_EPOCH) {
+                isFenced = true;
+                throw error.exception();
```
See comment above. We cannot generally raise exceptions from completion handlers.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,15 +261,77 @@ void run(long now) {
+        }
+
+        Node targetNode = null;
+        long expiryTime = now + requestTimeout;
```
Why do we expire at all? Seems we just want to keep retrying if we have a transactional request to send.

apurvam replied:
Good point. This code will be invoked repeatedly until the transactional request succeeds. So no need to have any expiry time.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,15 +261,77 @@ void run(long now) {
+        if (nextRequest.isEndTxnRequest() && transactionState.isCompletingTransaction() && accumulator.hasUnflushedBatches()) {
+            if (!accumulator.flushInProgress())
+                accumulator.beginFlush();
+            transactionState.didNotSend(nextRequest);
```
Maybe this should just be `reenqueue` or `retry`?

Commit:
Address a few more PR comments. Rebase on trunk / fix build and test failures

apurvam commented:
I fixed the correctness comments and fixed the merge conflicts after rebasing onto trunk.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,15 +261,77 @@ void run(long now) {
+        while (targetNode == null && nextIterationTime < expiryTime) {
+            try {
+                long timeRemaining = expiryTime - nextIterationTime;
+                if (nextRequest.needsCoordinator()) {
```
It's a bit awkward to handle finding the coordinator through the request queue in `TransactionState`. I would imagine something more like this:

```java
request = txnState.pollNextRequest();
coordinator = ensureCoordinatorReady(request);
send(request, coordinator);
```

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/ListOffsetRequest.java:
```diff
@@ -33,6 +33,7 @@
 import java.util.Set;
 
 public class ListOffsetRequest extends AbstractRequest {
+    public static final long LSO_TIMESTAMP = -3L;
```
This should be removed also.

hachikuji reviewed (commented):
A few comments on the transaction state machine and request handling.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -47,32 +190,166 @@ public boolean isValid() {
+                transactionCoordinator = null;
+                break;
+            default:
+                throw new IllegalStateException("Got an invalid coordintor type: " + type);
```
typo: "coordintor"

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -47,32 +190,166 @@ public boolean isValid() {
+        return inFlightRequestCorrelationId != NO_INFLIGHT_REQUEST_CORRELATION_ID;
+    }
+
+    // visible for testing
```
nit: if we only need it visible for testing, we can use default access.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -132,4 +414,332 @@ public synchronized void incrementSequenceNumber(TopicPartition topicPartition,
+            } else if (findCoordinatorResponse.error() == Errors.GROUP_AUTHORIZATION_FAILED) {
+                throw Errors.GROUP_AUTHORIZATION_FAILED.exception();
+            } else {
+                throw Errors.UNKNOWN.exception();
```
We're losing the error code from the response, which will definitely be useful if we ever happen to hit this case. In the consumer, we tend to do something like this:
```java
throw new KafkaException("Unexpected error in response: " + response.error().message());
```
Other response handlers have the same problem.

hachikuji reviewed (commented):
A few additional nitpicks.

Commit:
WIP

Commit:
Addressed PR comments. Main changes:

  1. Introduced more formal states to TransactionState and made the
state transition checks more formal.
  2. Stopped throwing exceptions from completion handlers.
  3. If the Transaction is in ERROR state, the next transactional method
will fail.
  4. The transaction can be in error state if any transactional request
or produce request within the transaction had a fatal error.

apurvam commented:
@hachikuji, I took care of all the major comments, and most of the minor ones. Main changes  include: 

1. Introduced more formal states to TransactionState and made the state transition checks more formal.
2. Stopped throwing exceptions from completion handlers.
3. If the Transaction is in ERROR state, the next transactional method will fail.
4. The transaction can be in error state if any transactional request or produce request within the transaction had a fatal error.

The only thing remaining is to cache errors from background requests like `AddPartitions` and `FindCoordinator` and propagate those on the next user-facing method. Currently we just log these errors. But that should be a minor change relatively speaking, so this is ready for review.

Commit:
Cached the last error in the transaction state so that if we invoke
another transactional method from an error state the cause can be
included in the 'IllegalStateException'.

Also moved around the methods in TransactionState so that they are more
logically grouped.

Commit:
Minor cleanup

apurvam commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -985,6 +1018,60 @@ private Record nextFetchedRecord() {
+            return abortedTransactions;
+        }
+
+        private boolean containsAbortMarker(RecordBatch batch) {
```
When the abort index patch comes in, we can leverage the `CONTROL_SEQUENCE` value to only look inside the batch there is a control message. This way we won't have to open all batches, potentially decompressing messages.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -951,7 +978,13 @@ private Record nextFetchedRecord() {
                     maybeEnsureValid(currentBatch);
+
+                    if (isolationLevel == IsolationLevel.READ_COMMITTED && isBatchAborted(currentBatch)) {
+                        continue;
```
When we skip a batch, we should ensure that `nextFetchOffset` is still updated. We can write a test case to assert the right offset position after receiving a record set where the abort marker is the last entry.

apurvam replied:
I added the code to update the offset of the skipped batches and added a test for this as well.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -985,6 +1018,60 @@ private Record nextFetchedRecord() {
+                    abortedProducerIds.add(batch.producerId());
+                    abortedTransactions.poll();
+                }
+                log.trace("Skipping aborted record batch with producerId {} and base offset {}", batch.producerId(), batch.baseOffset());
```
Mention the topic partition as well?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -985,6 +1018,60 @@ private Record nextFetchedRecord() {
+            Record firstRecord = batchIterator.hasNext() ? batchIterator.next() : null;
+            boolean containsAbortMarker = firstRecord != null && firstRecord.isControlRecord() && ControlRecordType.ABORT == ControlRecordType.parse(firstRecord.key());
+            if (containsAbortMarker && batchIterator.hasNext())
+                throw new CorruptRecordException("A record batch containing a control message contained more than one record.");
```
Mention topic partition and offset? Also, maybe `InvalidRecordException` is more appropriate?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -47,32 +231,93 @@ public boolean isValid() {
+    public synchronized void beginTransaction() {
+        ensureTransactional();
+        if (!transitionTo(State.IN_TRANSACTION))
+             throw new IllegalStateException("Producer isn't ready to begin a transaction.");
```
Maybe the message should say that the producer already has an on-going transaction?

apurvam commented:
System test run: http://jenkins.confluent.io/job/system-test-kafka-branch-builder/857/

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -523,6 +625,17 @@ private static int parseAcks(String acksString) {
+            if (transactionState.isTransactional() && !transactionState.hasPid())
+                throw new IllegalStateException("Cannot perform a 'send' before completing a call to initTransactions when transactions are enabled.");
+
+            if (transactionState.isFenced() && transactionState.isInTransaction())
```
Mentioned offline, but it seems we don't need the `isInTransaction` check here. No matter the state, if the producer is fenced, we should raise.

Commit:
Addressed more comments (online and offline):

  1. Changed the 'IllegalStateException' to 'KafkaException'
  2. Wake up the sender on every transactional request
  3. Tightened up the protection against errors. Once we are in an error
state, no transactional send or transactional rpc will be allowed until
abort is called.
  4. We only allow initTransactions from an uninitialized or error
state.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -253,35 +254,27 @@ public synchronized FutureTransactionalResult initializeTransactions() {
-
+        if (isFenced())
+            throw new ProducerFencedException("There is a newer producer using the same transactional.id.");
+        if (!transitionTo(State.ABORTING_TRANSACTION))
```
Do we need to reset `lastError` here? Seems like it's possible for the user to decide to abort on their own without first observing an awaited error.

hachikuji replied:
Never mind. As was explained offline, the last error is reset once the transaction is completed.

Commit:
Renamed TransactionState to TransactionManager.
Moved TransactionManager to the internals package.

Commit:
Throw an exception when a state transition fails detailing the failure.
We were previously logging the details and throwing a generic
exception, which was not very helpful.

Commit:
Fix build errors aftre rebase, miscellaneous cleanups

apurvam commented:
Rebased onto trunk and kicked off a new system test build with the latest changes: http://jenkins.confluent.io/job/system-test-kafka-branch-builder/863/

Commit:
Fix the visibility of various methods in the TransactionManager. Made a bunch package-private, since the class is now in the same package as the sender

Commit:
Handle the newly introduced CONCURRENT_TRANSACTIONS error code

Commit:
Minor improvements to the TransactionalManager

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -554,7 +664,13 @@ private static int parseAcks(String acksString) {
+            Callback interceptCallback = new InterceptorCallback<>(callback, this.interceptors, tp, transactionManager);
+
+            if (transactionManager != null && transactionManager.isInTransaction()) {
+                transactionManager.maybeAddPartitionToTransaction(new TopicPartition(record.topic(), record.partition()));
```
This breaks with an NPE when the user does not specify the partition explicitly in the `ProducerRecord`. We should use `tp` as the partition.

apurvam replied:
Good catch, fixed this.

Commit:
Add the existing topicPartition object to a transaction. Otherwise we won't add the partition if it isn't explicitly added to the record

apurvam commented:
System tests passed: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2017-04-26--001.1493284222--apurvam--exactly-once-transactional-clients--f8bdb71/report.html

Commit:
Minor cleanup

hachikuji reviewed (approved):
LGTM. Thanks for the patch! I'll merge once the most recent builds finish.

hachikuji commented:
Merging to trunk.
